### PR TITLE
WIP: EOS-20762: copy s3 buffer checksum to motr server

### PR DESCRIPTION
**Pull request by yeshpal-jain-seagate**
- copy checksum values passed from s3 in attr variable to motr buffers
- send checksum fops to io service

Ex:
S3:<seg0><seg1><seg2><seg3><seg0><seg1><seg2><seg3>
iofop1:
mo_buf {
b_nob = sizeof(b_addr)
b_addr = <seg0><seg3><seg4><seg7>
}

iofop2:
mo_buf {
b_nob = sizeof(b_addr)
b_addr = <seg1><seg2><seg5><seg6>
}

Signed-off-by: Yeshpal Jain <yeshpal.jain@seagate.com>

### CODE REVIEW

File: motr/pg.h [hunk 1]
```diff
@@ -803,6 +803,7 @@ struct target_ioreq {
 	 */
 	struct m0_bufvec               ti_bufvec;
 	struct m0_bufvec               ti_auxbufvec;
+	struct m0_bufvec               ti_attrbufvec;
```
Comment (sg-shankar) [inline motr/pg.h:806]: Please add comment describing what this structure will contain

File: motr/io_nw_xfer.c [hunk 2]
```diff
@@ -513,6 +517,8 @@ static void target_ioreq_seg_add(struct target_ioreq              *ti,
 	toff    = target_offset(frame, play, gob_offset);
 	pgstart = toff;
 	goff    = unit_type == M0_PUT_DATA ? gob_offset : 0;
+	coff    = di_cksum_offset(play, gob_offset);
```
Comment (sg-shankar) [inline motr/io_nw_xfer.c:520]: I think we can move this computation inside the while loop as goff changes in while loop
coff    = di_cksum_offset(play, goff);

File: motr/io_nw_xfer.c [hunk 3]
```diff
@@ -1078,6 +1133,15 @@ static int target_ioreq_init(struct target_ioreq    *ti,
 	if (ti->ti_bufvec.ov_buf == NULL)
 		goto fail;
 
+	ti->ti_attrbufvec.ov_vec.v_nr = nr;
```
Comment (sg-shankar) [inline motr/io_nw_xfer.c:1136]: Add TODO to optimize the size of buffer allocated

**Review by sg-shankar — CHANGES_REQUESTED** [review]
Code looks good, minor changes suggested

File: motr/io_nw_xfer.c [hunk 4]
```diff
@@ -736,6 +751,52 @@ static void irfop_fini(struct ioreq_fop *irfop)
...
+		return M0_ERR(-ENOMEM);
+	dst = dest->b_addr;
+	for (i = 0; i < src->ov_vec.v_nr; ++i) {
+		char str[64];
+		len_seg = src->ov_vec.v_count[i];
+		if (len_seg > 0) {
+			snprintf(str, len_seg, "%s", (char *)src->ov_buf[i]);
+			//YJC_TODO: length shoudl len_seg instead of len_seg -1
+			// this is only for debug because we have string 
+			memcpy(dst, src->ov_buf[i], len_seg - 1);
```
Comment (gshipra) [inline motr/io_nw_xfer.c:783]: memcpy to dst will wipe off previously written data.

dst + written_till_now something like this should happen.

**Pull request closed without merging**